File: tank_player_grandmother/src/TankLogic.cpp
/******************************************************************************
*
* Copyright Consoden AB, 2015
*
* Created by: Joel Ottosson / joot
*
*******************************************************************************/
#include "TankLogic.h"

#include "BfsHelper.h"
#include "GameMap.h"

const std::wstring TankLogic::PlayerName = L"grandmother"; //TODO: change to your team name
bool fired_last_round = false;

void TankLogic::MakeMove(Consoden::TankGame::GameStatePtr gameState)
{
    //-------------------------------------------------------
    //Example of a stupid tank logic:
    //Remove it and write your own brilliant version!
    //-------------------------------------------------------
    gm =  GameMap::GameMap(m_ownTankId, gameState);

    auto currentPosition=gm.OwnPosition();
    auto enemyPosition=gm.EnemyPosition();
    BfsHelper bfs(gameState, currentPosition);
    Consoden::TankGame::Direction::Enumeration moveDirection = Consoden::TankGame::Direction::Neutral;
    Consoden::TankGame::Direction::Enumeration towerDirection;

    //check if enemy is at same line as we are.
    bool fire = false;
    towerDirection = isInSight();
    if(towerDirection != Consoden::TankGame::Direction::Neutral && !fired_last_round){
		fire = true;
		fired_last_round = true;

    }else{
    	fired_last_round = false;
    }


    if(!fire){
		if (bfs.CanReachSquare(enemyPosition)) {
			// It is possible to move all the way to the enemy, do it
			moveDirection=bfs.FindDirection(currentPosition, bfs.BacktrackFromSquare(enemyPosition));

		} else {
			//Find an empty sqaure we can move to, otherwise move downwards
			moveDirection=Consoden::TankGame::Direction::Neutral;
			if (!gm.IsWall(gm.Move(currentPosition, Consoden::TankGame::Direction::Left)) &&
				!gm.IsMine(gm.Move(currentPosition, Consoden::TankGame::Direction::Left)))
			{
				moveDirection=Consoden::TankGame::Direction::Left;
			}
			else if (!gm.IsWall(gm.Move(currentPosition, Consoden::TankGame::Direction::Right)) &&
					 !gm.IsMine(gm.Move(currentPosition, Consoden::TankGame::Direction::Right)))
			{
				moveDirection=Consoden::TankGame::Direction::Right;
			}
			else if (!gm.IsWall(gm.Move(currentPosition, Consoden::TankGame::Direction::Up)) &&
					 !gm.IsMine(gm.Move(currentPosition, Consoden::TankGame::Direction::Up)))
			{
				moveDirection=Consoden::TankGame::Direction::Up;
			}
			else if (!gm.IsWall(gm.Move(currentPosition, Consoden::TankGame::Direction::Down)) &&
					 !gm.IsMine(gm.Move(currentPosition, Consoden::TankGame::Direction::Down)))
			{
				moveDirection=Consoden::TankGame::Direction::Down;
			}
		}
    }

    //Move our joystick.
    SetJoystick(moveDirection, towerDirection, fire, false);
}


Consoden::TankGame::Direction::Enumeration TankLogic::isInSight(){

	auto own = gm.OwnPosition();
	auto enemy = gm.EnemyPosition();
	int xdiff = own.first - enemy.first;
	int ydiff = own.second - enemy.second;

	if(xdiff == 0){
		if(ydiff > 0){
			return Consoden::TankGame::Direction::Up;
		}else{
			return Consoden::TankGame::Direction::Down;
		}

	}else if(ydiff == 0){
		if(xdiff > 0){
			return Consoden::TankGame::Direction::Left;
		}else{
			return Consoden::TankGame::Direction::Right;
		}
	}else{
		return Consoden::TankGame::Direction::Neutral;
	}

}

File: src/players/tank_player_templates/tank_player_java/src/cwg/TankLogic.java
/******************************************************************************
*
* Copyright Consoden AB, 2014
*
* Created by: Joel Ottosson / joot
*
******************************************************************************/
package cwg;


class TankLogic {
	
	public final static String PLAYER_NAME = "tank_player_java"; //TODO: Change to your team name
	
	public interface JoystickHandler {
		public void setJoystick(consoden.tankgame.Direction moveDirection,
								consoden.tankgame.Direction towerDirection,
								boolean fire, boolean fireLaser, boolean deploySmoke,
								boolean fireRedeemer, int redeemerTimer);
	}
		
	private int tankId;
	private JoystickHandler joystickHandler;
	
	public TankLogic(int tankId, JoystickHandler joystickHandler){
		this.tankId=tankId;
		this.joystickHandler=joystickHandler;
	}
	
	public void makeMove(consoden.tankgame.GameState gameState) {
		
	    //TODO: implement your own tank logic and call setJoystick

		//-------------------------------------------------------
		//Example of a stupid tank logic:
		//Remove it and write your own brilliant version!
		//-------------------------------------------------------
	    GameMap gm=new GameMap(tankId, gameState); //helper object
	    Bfs bfs=new Bfs(gameState, gm.getOwnPosition(), true);	    
	    
		Position currentPosition = gm.getOwnPosition(); //this is our current tank position
		Position enemyPosition = gm.getEnemyPosition(); //this is the enemy tank position
		
		consoden.tankgame.Direction moveDirection = consoden.tankgame.Direction.NEUTRAL;

		if (bfs.canReachSquare(enemyPosition)) { //if we can reach the enemy, get him			
			moveDirection=bfs.backtrackFromSquare(enemyPosition);
			System.out.println("Can reach enemy!");
		}
		else { //find any empty square			
			if (!gm.isWall(gm.move(currentPosition, consoden.tankgame.Direction.LEFT)) &&
				!gm.isMine(gm.move(currentPosition, consoden.tankgame.Direction.LEFT))) {
				moveDirection=consoden.tankgame.Direction.LEFT;
			}
			else if (!gm.isWall(gm.move(currentPosition, consoden.tankgame.Direction.RIGHT)) &&
					 !gm.isMine(gm.move(currentPosition, consoden.tankgame.Direction.RIGHT))) {
				moveDirection=consoden.tankgame.Direction.RIGHT;
			}
			else if (!gm.isWall(gm.move(currentPosition, consoden.tankgame.Direction.UP)) &&
					 !gm.isMine(gm.move(currentPosition, consoden.tankgame.Direction.UP))) {
				moveDirection=consoden.tankgame.Direction.UP;
			}
			else if (!gm.isWall(gm.move(currentPosition, consoden.tankgame.Direction.DOWN)) &&
					 !gm.isMine(gm.move(currentPosition, consoden.tankgame.Direction.DOWN))) {
				moveDirection=consoden.tankgame.Direction.DOWN;
			}
		}

		//Advanced tower aim stategy
		consoden.tankgame.Direction towerDirection = 
				consoden.tankgame.Direction.values()[(1 + currentPosition.x + currentPosition.y) % 4];			

		//Of course we always want to fire
		boolean fire = true;
		
	    boolean fireLaser = false;
	    //if we can we fire lasers!!!
	    if(gm.getLaserCount() > 0){
	    	fireLaser = true;
	    }
	    
	    if (gm.isPenguinAlive()) {
        	System.out.println("Penguin is alive");
    	} else {        
        	System.out.println("Penguin is dead");
    	}
	    
		setJoystick(moveDirection, towerDirection, fire, fireLaser);
	}
	
	public void setJoystick(consoden.tankgame.Direction moveDirection,
								consoden.tankgame.Direction towerDirection,
								boolean fire, boolean fireLaser)
	{
		//Move our joystick.
		joystickHandler.setJoystick(moveDirection, towerDirection, fire, fireLaser, false, false, 0);
	}
}


